src/TesApi.Web/BatchScheduler.cs

```diff
@@ -573,6 +526,70 @@ static bool IsPoolQuotaException(string code)
+                        logger.LogWarning($"TES task: {tesTask.Id} AzureBatchQuotaMaxedOutException.Message: {azureBatchQuotaMaxedOutException.Message} . Not enough quota available. Task will remain with state QUEUED.");
+                        neededPools.Add(poolName);
+                        break;
+
+                    case AzureBatchLowQuotaException azureBatchLowQuotaException:
+                        tesTask.State = TesState.SYSTEMERROREnum;
+                        tesTask.AddTesTaskLog(); // Adding new log here because this exception is thrown from CheckBatchAccountQuotas() and AddTesTaskLog() above is called after that. This way each attempt will have its own log entry.
+                        tesTask.SetFailureReason("InsufficientBatchQuota", azureBatchLowQuotaException.Message);
+                        logger.LogError(azureBatchLowQuotaException, $"TES task: {tesTask.Id} AzureBatchLowQuotaException.Message: {azureBatchLowQuotaException.Message}");
+                        break;
+
+                    case AzureBatchVirtualMachineAvailabilityException azureBatchVirtualMachineAvailabilityException:
+                        tesTask.State = TesState.SYSTEMERROREnum;
+                        tesTask.AddTesTaskLog(); // Adding new log here because this exception is thrown from GetVmSizeAsync() and AddTesTaskLog() above is called after that. This way each attempt will have its own log entry.
+                        tesTask.SetFailureReason("NoVmSizeAvailable", azureBatchVirtualMachineAvailabilityException.Message);
+                        logger.LogError(azureBatchVirtualMachineAvailabilityException, $"TES task: {tesTask.Id} AzureBatchVirtualMachineAvailabilityException.Message: {azureBatchVirtualMachineAvailabilityException.Message}");
+                        break;
+
+                    case TesException tesException:
+                        tesTask.State = TesState.SYSTEMERROREnum;
+                        tesTask.SetFailureReason(tesException);
+                        logger.LogError(tesException, $"TES task: {tesTask.Id} TesException.Message: {tesException.Message}");
+                        break;
+
+                    case BatchClientException batchClientException:
+                        tesTask.State = TesState.SYSTEMERROREnum;
+                        tesTask.SetFailureReason("BatchClientException", string.Join(",", batchClientException.Data.Values), batchClientException.Message, batchClientException.StackTrace);
+                        logger.LogError(batchClientException, $"TES task: {tesTask.Id} BatchClientException.Message: {batchClientException.Message} {string.Join(",", batchClientException?.Data?.Values)}");
+                        break;
+
+                    case BatchException batchException when batchException.InnerException is Microsoft.Azure.Batch.Protocol.Models.BatchErrorException batchErrorException && IsJobQuotaException(batchErrorException.Body.Code):
```

Comment:
I believe these exceptions should be logged. It is hard to troubleshoot root causes if the root exception information is lost.

Reply:
Same as above. I will, however, test to see what shows up in the logs.

Reply:
I double checked. In `Log Analytics Workspace`, under `LogManagement`, in addition to `AppTraces`, there's `AppExceptions`. Inside of that, the entire exception chain (following inner exceptions) including stack traces is present for every log where an exception was given to the API, as far as I can see. It's wonderful for root cause analysis.